### DISCUSSION TITLE
Input handling library

### DISCUSSION THREAD

**Discussion by espkk** [storm-devs/storm-engine#19] [Ideas]
Candidates:
1. gainput
2. SDL2
3. your option

Please, share your options, your analysis of the pros and cons

**Comment by q4a** [comment]
We need input handling, window handing, audio handling and may be more. And this things can be done in SDL2.
I have some experience with SDL. It can help with File System part too:
https://wiki.libsdl.org/SDL_GetBasePath - was used in my previous projects to save logs and save files.

SDL2 pros:
+ support and development
+ wide library will reduce the number of dependencies
+ (my personal) I have some experience with SDL

SDL2 cons:
I heard, that SDL2 has some performance problem on Windows. My friends from OpenXRay got lower FPS when switching mouse input handling from native Windows to SDL2 (2.0.5). But OpenXRay is First-person Shooter and they need to have big FPS, they use a bit old SDL2 (there is 2.0.14). They are thinking about simple `#ifdef` some code to using the old version for Windows and SDL2 for the rest (including the Nintendo switch port).

**Comment by q4a** [comment]
gainput has big cons - it was not updated since march 2018 and looks like abandoned:
https://github.com/jkuhlmann/gainput/issues/77

**Comment by np-c0mp1ete** [comment]
I think it's better to stick to maintained solutions. But also in case we later realize it was a mistake, I'd suggest introducing some abstraction layer on top of a library we choose to be able to easily replace it if needed.

**Comment by np-c0mp1ete** [comment]
@Hammie, @VictorJL, what do you think about the topic?

**Reply by Hammie** [reply]
I only have some experience with SDL, which I was happy about, so I don't have strong opinions on this.

Because mouse & keyboard input is often tied closely to window/display management, I like that SDL can be used for both.

I'm not sure how much use gainput has seen, so I don't know how stable it is, but I would still prefer using a no-longer maintained library over creating a new library. We would have to maintain it ourselves in either case anyway.

An abstraction layer also seems like a good idea to me, and probably required to get key-rebinding and configuring to work well.
